### PR TITLE
[coverage] Fixes and updates

### PR DESCRIPTION
**Pull request by alphan**
This PR addresses some of tasks captured in #16761:

* Adds the `measure_on_target_coverage` config setting to link with the correct dependencies for coverage runtime,
* Fixes `sw/device/lib/testing/test_framework/ottf_common.ld` to keep the symbols in `.init_array.*`,
* Improve coverage buffer printing, and
* Updates llvm_clang_rt_profile to `llvmorg-13.0.1`.

### CODE REVIEW

**Comment by alphan** [comment]
The last commit reduces the remote timeout from 60 to 5 per @milesdai recommendation as a temporary/speculative workaround.

File: sw/device/lib/testing/test_framework/coverage_llvm.c
Comment (dmcardle) [inline sw/device/lib/testing/test_framework/coverage_llvm.c:60]: A few thoughts:

* It looks like we're hexdumping the buffer in reverse order — why is that?
* Do we have any sort of `span` or `buffer` type?
* It might be nice to write a general-purpose hexdump (although it's probably better not to prematurely generalize).

File: util/device_profile_data.py
Comment (dmcardle) [inline util/device_profile_data.py:54]: The commit message says the output is big-endian — is this parameter incorrect? The `print.h` documentation says `%!x` hexdumps in big-endian order.

**Review by dmcardle — CHANGES_REQUESTED** [review]
Great work, @alphan! I've left you entirely too many comments, only a few real requests for changes. Thanks!

**Review by dmcardle — APPROVED** [review]
Just learned that "request changes" can actually block the PR up. None of my comments are so critical that they need to block the PR, so I'm upgrading to "Approve" (modulo nits).

**Comment by alphan** [comment]
> Just learned that "request changes" can actually block the PR up. None of my comments are so critical that they need to block the PR, so I'm upgrading to "Approve" (modulo nits).

Thanks for the quick review @dmcardle !

File: util/device_profile_data.py
Comment (dmcardle) [reply util/device_profile_data.py:54]: Ignore this. As we discussed offline, I see now that `int` is doing the parsing, and `byteorder='little'` is how we're generating generating the byte array.

File: sw/device/BUILD
Comment (alphan) [reply sw/device/BUILD:24]: TIL, done.

File: sw/device/lib/base/BUILD
Comment (alphan) [reply sw/device/lib/base/BUILD:33]: done

File: sw/device/lib/testing/test_framework/coverage_llvm.c
Comment (alphan) [reply sw/device/lib/testing/test_framework/coverage_llvm.c:36]: You're right even though I doubt that we will ever hit this, the right thing to do here is to compare the `uint64_t` llvm profile buffer size with the size of the `static` `buf`.

File: sw/device/lib/testing/test_framework/coverage_llvm.c
Comment (alphan) [reply sw/device/lib/testing/test_framework/coverage_llvm.c:43]: Since we already check above and we don't support the `z` specifier the best option is to cast. Done.

File: util/device_profile_data.py
Comment (alphan) [reply util/device_profile_data.py:47]: No because we remove the new lines in the line just above this one for a simpler regexp.

There is also [this](https://cs.opensource.google/opentitan/opentitan/+/master:sw/host/opentitanlib/src/uart/console.rs;drc=3cc6786654a3bb4e5f6e81dc45b7c9aee34515eb;l=90).

File: sw/device/lib/testing/test_framework/coverage_llvm.c
Comment (dmcardle) [inline sw/device/lib/testing/test_framework/coverage_llvm.c:54]: Oh closer inspection, I think lines 53 and 54 should use `bytes_per_line` instead of `kBytesFirstLine`.

File: util/device_profile_data.py
Comment (alphan) [reply util/device_profile_data.py:36]: done

File: sw/device/lib/testing/test_framework/coverage_llvm.c
Comment (alphan) [reply sw/device/lib/testing/test_framework/coverage_llvm.c]: Good catch! Fixed. LLVM profile buffers always end up larger than ~100 bytes even for really small test programs but regardless we must handle this case correctly.

File: util/device_profile_data_test.py
Comment (alphan) [reply util/device_profile_data_test.py]: Also replaced the `EOT` with `\x04`.

File: sw/device/lib/base/BUILD
Comment (milesdai) [inline sw/device/lib/base/BUILD:33]: Nit: change the first commit message to refer to `math_polyfills` instead of `math_builtins_no_builtins`.

File: sw/device/lib/testing/test_framework/coverage_llvm.c
Comment (milesdai) [inline sw/device/lib/testing/test_framework/coverage_llvm.c:39]: `(uint32_t)buf_size` --> `(uint32_t)buf_size_u64`?

File: sw/device/lib/testing/test_framework/ottf_main.c
Comment (moidx) [inline sw/device/lib/testing/test_framework/ottf_main.c:103]: Just to double check. Are we okay with doing this in DV? Not sure if this will trigger timeouts in the nightly regression.

File: util/BUILD
Comment (moidx) [inline util/BUILD:82]: Do we run this in CI?

**Review by moidx — APPROVED** [review]

File: util/BUILD
Comment (milesdai) [reply util/BUILD:82]: [Yes](https://dev.azure.com/lowrisc/opentitan/_build/results?buildId=105310&view=logs&j=5df5adf8-3577-5927-f0b2-31afed2697ee&t=1253ba27-4e8c-57af-d0ba-888237644875&l=3479)

The Earl Gray SW Build and Test job should pick up any stray tests in the repo that aren't part of the "device" testing (e.g. functests and e2e tests)

File: sw/device/lib/testing/test_framework/ottf_main.c
Comment (alphan) [reply sw/device/lib/testing/test_framework/ottf_main.c:103]: Thanks for checking @moidx! `coverage_send_buffer` is NOOP by default. We have the actual implementation only if the config setting is passed in.

File: ci/.bazelrc
Comment (milesdai) [inline ci/.bazelrc:24]: We can indicate that this change is not temporary in the commit message.

**Review by milesdai — APPROVED** [review]

File: sw/device/lib/testing/test_framework/coverage_llvm.c
Comment (alphan) [reply sw/device/lib/testing/test_framework/coverage_llvm.c:39]: Ah, thanks for catching this, I forgot that I commented out the build part from my local script :)

**Comment by alphan** [comment]
Merging now because the previous CI run passed and the last push only includes a compile fix for a file that's not compiled in CI yet.

**Comment by alphan** [comment]
Thanks for the review @milesdai !

**Pull request merged**
